src/model/offset.ts

```diff
@@ -0,0 +1,64 @@
+import { strict as assert } from "assert";
+import { Timestamp } from "./timestamp";
+
+export class Offset {
+  public static readonly zero = new Offset("ZERO", BigInt(1), Timestamp.zero);
+
+  public constructor(
+    public readonly id: string,
+    public readonly height: bigint,
+    public readonly timestamp: Timestamp
+  ) {
+    assert(id.length > 0 || height == BigInt(0));
+  }
+
+  public equals(offset: Offset): boolean {
+    return this.id == offset.id;
+  }
+
+  public sameHeight(offset: Offset): boolean {
+    return this.height == offset.height;
+  }
+
+  public canSucceed(offset: Offset): boolean {
+    return (
+      this.height - BigInt(1) == offset.height &&
+      this.timestamp.greaterThan(offset.timestamp)
+    );
+  }
+
+  public canPrecede(offset: Offset): boolean {
+    if (this.equals(Offset.zero)) return true;
+
+    return (
+      this.height + BigInt(1) == offset.height &&
+      this.timestamp.lessThan(offset.timestamp)
+    );
+  }
+
+  public dump(): string {
```

Comment:
If it's a function which formats the string which can be parsed with `parse`, we should come up with a better name. I think `dump` is supposed to be used just for formatting in logs. How about `format` or `toString`?